grype/pkg/cpe_provider.go

```diff
@@ -0,0 +1,100 @@
+package pkg
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/anchore/syft/syft/cpe"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/sbom"
+	"github.com/anchore/syft/syft/source"
+)
+
+const cpeInputPrefix = "cpe:"
+
+type CPELiteralMetadata struct {
+	CPE string
+}
+
+func cpeProvider(userInput string) ([]Package, Context, *sbom.SBOM, error) {
+	reader, ctx, err := getCPEReader(userInput)
+	if err != nil {
+		return nil, Context{}, nil, err
+	}
+
+	return decodeCPEFile(reader, ctx)
+}
+
+func getCPEReader(userInput string) (r io.Reader, ctx Context, err error) {
+	if strings.HasPrefix(userInput, cpeInputPrefix) {
+		ctx.Source = &source.Description{
+			Metadata: CPELiteralMetadata{
+				CPE: userInput,
+			},
+		}
+		return strings.NewReader(userInput), ctx, nil
+	}
+	return nil, ctx, errDoesNotProvide
+}
+
+func decodeCPEFile(reader io.Reader, ctx Context) ([]Package, Context, *sbom.SBOM, error) {
+	scanner := bufio.NewScanner(reader)
+	var packages []Package
+	var syftPkgs []pkg.Package
+
+	for scanner.Scan() {
+		rawLine := scanner.Text()
+		p, syftPkg, err := cpeToPackage(rawLine)
+		if err != nil {
+			return nil, Context{}, nil, err
+		}
+
+		if p != nil {
+			packages = append(packages, *p)
+		}
+		if syftPkg != nil {
+			syftPkgs = append(syftPkgs, *syftPkg)
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
```

Comment:
TIL `scanner.Err()` returns nil if the only err is `io.EOF`: https://github.com/golang/go/blob/64d82cd72c222caa93b2f71c5970a00ec4e7929a/src/bufio/scan.go#L99